Add UEFI and TF-A as certificates for LmP

angolini opened the pull request:
While on this reorganize and fix some minor stuff.

It's still missing:

- [ ] which is the "better looking" diagram? with a legend or with the variable name near each edge?
- [ ] a section describing how to handle tfa certificate files (such as rotation)

ricardosalveti commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -9,6 +9,103 @@ modules. The device RoT key (the key used for secure boot, for example) is not
 listed here as it is not an online key and is not used during the
 FoundriesFactory build.
 
+Security Boot Flow
```
I think we can probably just say Secure Boot Flow instead.

ricardosalveti commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -9,6 +9,103 @@ modules. The device RoT key (the key used for secure boot, for example) is not
+------------------
+
+The security boot flow starts from the boot rom and the first certificate
+is commounly from the fuses. After that step, the flow is defined by the SoC
```
I would drop 'and the first certificate is commounly from the fuses.'

ricardosalveti commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -9,6 +9,103 @@ modules. The device RoT key (the key used for secure boot, for example) is not
+
+   digraph {
+        graph [
+            label = "Security boot flow for i.MX machines"
```
Secure boot flow

ricardosalveti commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -9,6 +9,103 @@ modules. The device RoT key (the key used for secure boot, for example) is not
+        "Boot ROM"     -> "SPL"                  [label = "1"];
+        "SPL"          -> "uboot.itb"            [label = "2"];
+        "uboot.itb"    -> "bootscr"              [label = "2"];
+        "uboot.itb"    -> "TF-A"                 [label = "2"];
```
TF-A is optional on some targets, like imx6/7/7ulp (most 32 bits imx variants).

ricardosalveti commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -9,6 +9,103 @@ modules. The device RoT key (the key used for secure boot, for example) is not
+        "uboot.itb"    -> "bootscr"              [label = "2"];
+        "uboot.itb"    -> "TF-A"                 [label = "2"];
+        "uboot.itb"    -> "OP-TEE"               [label = "3"];
+        "uboot.itb"    -> "uboot proper"         [label = "2"];
```
U-Boot proper

ricardosalveti commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -9,6 +9,103 @@ modules. The device RoT key (the key used for secure boot, for example) is not
+        "uboot.itb"    -> "TF-A"                 [label = "2"];
+        "uboot.itb"    -> "OP-TEE"               [label = "3"];
+        "uboot.itb"    -> "uboot proper"         [label = "2"];
+        "uboot proper" -> "fitImage"             [label = "4"];
```
To avoid confusion we should probably say kernel fitImage (as u-boot.itb is also a fitimage).

ricardosalveti commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -9,6 +9,103 @@ modules. The device RoT key (the key used for secure boot, for example) is not
+                key [label=<<table border="0" cellpadding="2" cellspacing="0" cellborder="0">
+                        <tr><td align="right" port="i1">1 - fuse </td></tr>
+                        <tr><td align="right" port="i2">2 - UBOOT_SPL_SIGN_KEYNAME</td></tr>
+                        <tr><td align="right" port="i2">3 - OPTEE_TA_SIGN_KEY</td></tr>
```
TA sign key can be added here similarly as done for kernel modules, as it is used to validate the trusted applications loaded at runtime (TAs).

ricardosalveti commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -9,6 +9,103 @@ modules. The device RoT key (the key used for secure boot, for example) is not
+
+   }
+
+The next diagram shows the security boot flow for stm32 based machines:
```
STM32MP15

ricardosalveti commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -9,6 +9,103 @@ modules. The device RoT key (the key used for secure boot, for example) is not
+                arrowhead=none
+        ];
+        "Boot ROM"     -> "TF-A"                 [label = "fuse"];
+        "TF-A"         -> "uboot.itb"            [label = "UBOOT_SIGN_KEYNAME"];
```
uboot.itb is not used here, instead TF-A validates fip.bin (which is a container with optee and u-boot proper in it).

ricardosalveti replied:
And with bootscr.

vanmaegima commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -9,6 +9,90 @@ modules. The device RoT key (the key used for secure boot, for example) is not
+
+The secure boot flow starts from the boot rom. After that step, the flow is
+defined by the SoC vendor as it depends on the architecture,
+and also the BSP provided by the SoC vendor.
```
```suggestion
and also on the BSP provided by the SoC vendor.
```

vanmaegima commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -56,8 +142,10 @@ The directory structure shown below:
         │   │   ├── privkey_modsign.pem
         │   │   ├── spldev.crt
         │   │   ├── spldev.key
+        │   │   ├── tf-a
         │   │   ├── ubootdev.crt
         │   │   ├── ubootdev.key
+        │   │   ├── uefi
```
Should we say that a factory will not include all the files listed here as that is SoC dependent?

angolini replied:
The generated keys for the factory do not depend on SoC. At least, my latest factory created for imx machine does include uefi and tf-a.

vanmaegima commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -43,6 +127,8 @@ The file ``x509.genkey`` is a configuration file used for creating
 format, and is used for signing Linux Kernel Modules. This is used by
 configuring the variable ``MODSIGN_PRIVKEY``.
 
+The **UEFI** certificates are detailed in :ref:`ref-secure-boot-uefi`.
```
Should we list the UEFI certificates briefly here for consistency?

vanmaegima commented:
I'd love to review this again after the page is building so I can check the graphs.

ricardosalveti commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -4,10 +4,99 @@ Crypto Keys Used by FoundriesFactory at Build Time
+Secure Boot Flow
+------------------
+
+The secure boot flow starts from the boot rom. After that step, the flow is
```
rom is usually all caps, like boot ROM, so we might want to change our references here.

ricardosalveti commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -4,10 +4,99 @@ Crypto Keys Used by FoundriesFactory at Build Time
+LmP implements three variations of boot flow, using Secondary Program Loader (SPL),
+TF-A or Unified Extensible Firmware Interface (UEFI).
+
+Some boot flow also includes OP-TEE, .
```
Extra comma.

ricardosalveti commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -4,10 +4,99 @@ Crypto Keys Used by FoundriesFactory at Build Time
+        "Boot ROM"        -> "SPL"                  [label = "fuse"];
+        "SPL"             -> "uboot.itb"            [label = "UBOOT_SPL_SIGN_KEYNAME"];
+        "uboot.itb"       -> "bootscr"              [label = "UBOOT_SPL_SIGN_KEYNAME"];
+        "uboot.itb"       -> "TF-A (optional)"      [label = "TF_A_SIGN_KEY_PATH"];
```
I guess instead of optional just say (arm64 devices)?

In theory it can be also enabled on 32-bits devices (like done for stm32), but this is not supported by NXP.

ricardosalveti commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -4,10 +4,99 @@ Crypto Keys Used by FoundriesFactory at Build Time
+	edge [
+                arrowhead=none
+        ];
+        "Boot ROM"              -> "fip.bin (TF-A and OP-TEE)"[label = "fuse"];
```
I guess since we're talking about Intel (and would apply to other targets as well), we can just say boot rom -> UEFI as TF-A / OP-TEE are not available on intel and not necessarily available on ARM either.

ricardosalveti commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -4,10 +4,99 @@ Crypto Keys Used by FoundriesFactory at Build Time
+                arrowhead=none
+        ];
+        "Boot ROM"              -> "fip.bin (TF-A and OP-TEE)"[label = "fuse"];
+        "fip.bin (TF-A and OP-TEE)" -> "uefi"                 [label = "TF_A_SIGN_KEY_PATH"];
```
This is atm only used with stm32, so fine to drop and have rom -> uefi.

ricardosalveti commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -4,10 +4,99 @@ Crypto Keys Used by FoundriesFactory at Build Time
+        ];
+        "Boot ROM"              -> "fip.bin (TF-A and OP-TEE)"[label = "fuse"];
+        "fip.bin (TF-A and OP-TEE)" -> "uefi"                 [label = "TF_A_SIGN_KEY_PATH"];
+        "uefi"                  -> "systemd.boot"         [label = "UEFI_SIGN_KEYDIR"];
```
Probably good to use all caps with UEFI.

For systemd boot, write as systemd-boot.

ricardosalveti commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -4,10 +4,99 @@ Crypto Keys Used by FoundriesFactory at Build Time
+        "Boot ROM"              -> "fip.bin (TF-A and OP-TEE)"[label = "fuse"];
+        "fip.bin (TF-A and OP-TEE)" -> "uefi"                 [label = "TF_A_SIGN_KEY_PATH"];
+        "uefi"                  -> "systemd.boot"         [label = "UEFI_SIGN_KEYDIR"];
+        "systemd.boot"          -> "Linux kernel"         [label = "TF_A_SIGN_KEY_PATH"];
```
meta-lmp-base/recipes-core/systemd/systemd-boot_%.bbappend

```
                for efi in `find ${B}/src/boot/efi -name '*.efi'`; do
                        sbsign --key ${UEFI_SIGN_KEYDIR}/DB.key --cert ${UEFI_SIGN_KEYDIR}/DB.crt $efi
                        sbverify --cert ${UEFI_SIGN_KEYDIR}/DB.crt $efi.signed
                        mv $efi.signed $efi
                done
```
It uses DB.key for signing EFI payloads. Same also done for kernel, check meta-lmp-base/classes/kernel-lmp-efi.bbclass.

ricardosalveti commented on source/reference-manual/security/ff-security-summary.rst:
```diff
@@ -53,22 +59,33 @@ The exact list of keys used for the boot stack depends on the hardware used. Som
+   * - UEFI :ref:`(detail) <ref-secure-boot-uefi>`
+     - RSA 2048 or SHA 256
+     - Owned by the customer, available as an online key for FoundriesFactory CI
+     - ``UEFI_SIGN_KEYDIR``
```
Here we just configure the keydir, but the actual key used is DB.key.

doanac commented:
Docs for c55629dfaf7fca40f724bb01627a4633e07f484a are browsable at: https://ci.foundries.io/projects/fio-docs/builds/2098/docs/artifacts/html/index.html

doanac commented:
Docs for fa27813078c8fe1e93f47b6d13c3dccaad6586a7 are browsable at: https://ci.foundries.io/projects/fio-docs/builds/2099/docs/artifacts/html/index.html

doanac commented:
Docs for 12146c7f341bfe31c05ef4867d26e41d7b71bedb are browsable at: https://ci.foundries.io/projects/fio-docs/builds/2101/docs/artifacts/html/index.html

doanac commented:
Docs for f49c4ed1f81cfbed0203e54b57a991ddefe27a41 are browsable at: https://ci.foundries.io/projects/fio-docs/builds/2104/docs/artifacts/html/index.html

doanac commented:
Docs for 346d4e6763a5655fad5d620af14fa9feb36f1a1b are browsable at: https://ci.foundries.io/projects/fio-docs/builds/2105/docs/artifacts/html/index.html

doanac commented:
Docs for d68ab030370b97aef3288b4340af9fc3c339293a are browsable at: https://ci.foundries.io/projects/fio-docs/builds/2106/docs/artifacts/html/index.html

doanac commented:
Docs for 93a30bfff1ce86c3005e3abfe8e559bbddb3dbe4 are browsable at: https://ci.foundries.io/projects/fio-docs/builds/2107/docs/artifacts/html/index.html

kprosise commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -3,47 +3,144 @@
+------------------
+
+The secure boot flow starts from the boot ROM. After that step, the flow is
+defined by the SoC vendor as it depends on the architecture and BSP provided by the SoC vendor.
```
```suggestion
defined by the SoC vendor as it depends on the architecture and the BSP they provide.
```
Just shortening it a little, as SoC vendor is already mentioned.

kprosise commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -3,47 +3,144 @@
+defined by the SoC vendor as it depends on the architecture and BSP provided by the SoC vendor.
+
+LmP implements three variations of boot flow, using Secondary Program Loader (SPL),
+TF-A or Unified Extensible Firmware Interface (UEFI).
```
```suggestion
TF-A, or Unified Extensible Firmware Interface (UEFI).
```

kprosise commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -3,47 +3,144 @@
+   }
+
+
+When a FoundriesFactory is created, by default, two sets of keys are created under
```
```suggestion
When a Factory is created, by default, two sets of keys are created under
```

kprosise commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -3,47 +3,144 @@
 ``lmp-manifest`` repository:
 
 * ``conf/keys``: The key set is a copy of the default LmP public keys.
 * ``factory-keys``: The key set is created during the FoundriesFactory creation
```
```suggestion
* ``factory-keys``: The key set is created during the Factory's creation
```

kprosise commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -67,29 +166,33 @@ The directory structure shown below:
 ---------------------------------------
 
-Each FoundriesFactory is created with a unique key set, however it is highly
+Each FoundriesFactory is created with a unique key set. However, it is highly
```
```suggestion
Each Factory is created with a unique key set. However, it is highly
```

kprosise commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -165,7 +292,7 @@ as shown in the following command:
 
 .. tip::
-        Don't forget to push the new keys to get it included in the next CI
+        Remind to push the new keys to get it included in the next CI
```
```suggestion
        Remember to push the new keys to get it included in the next CI
```

kprosise commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -122,14 +228,35 @@ OP-TEE keys
+            -keyout privkey_ec_prime256v1.pem
+
+.. tip::
+        Remind to push the new keys to get it included in the next CI
```
```suggestion
        Remember to push the new keys to get it included in the next CI
```

kprosise commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -67,29 +166,33 @@ The directory structure shown below:
 
-In the next sections, the command line on how to create the key pair for U-Boot,
+In the following sections, the command line on how to create the key pair for U-Boot,
 OP-TEE and Linux Kernel Modules. Assuming the ``lmp-manifest`` repository is
```
```suggestion
OP-TEE and Linux Kernel Modules is shown. Assuming the ``lmp-manifest`` repository is
```

kprosise reviewed (commented):
Left some suggestions. Overall it looks great!

doanac commented:
Docs for 58abc0b42e4f5a8304beb781ff2687c9cb9e6655 are browsable at: https://ci.foundries.io/projects/fio-docs/builds/2108/docs/artifacts/html/index.html

doanac commented:
Docs for 55aad2a0e478e52076f3ea853bff54b8b85b5604 are browsable at: https://ci.foundries.io/projects/fio-docs/builds/2109/docs/artifacts/html/index.html

doanac commented:
Docs for a7f619b74b45081fec5ba930ed684cc549458e27 are browsable at: https://ci.foundries.io/projects/fio-docs/builds/2110/docs/artifacts/html/index.html

doanac commented:
Docs for 96571d748e8a43562613b249dce8f92281a16831 are browsable at: https://ci.foundries.io/projects/fio-docs/builds/2111/docs/artifacts/html/index.html

angolini commented:
@ricardosalveti @vanmaegima @Tim-Anderson @kprosise The diagram is finally showing the letters. Could you, please, review again?

vanmaegima commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -3,47 +3,141 @@
-FoundriesFactory build.
-
-When a FoundriesFactory is created, by default two set of keys are created under
+The LmP build system can use some online keys to sign some boot components of
```
Suggestion: not required to implement.
```suggestion
By default, the LmP build system uses online keys to sign some boot components of
```

vanmaegima commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -3,47 +3,141 @@
+The secure boot flow starts from the boot ROM. After that step, the flow is
+defined by the SoC vendor as it depends on the architecture and the BSP they provide.
+
+LmP implements three variations of boot flow, using Secondary Program Loader (SPL),
+TF-A, or Unified Extensible Firmware Interface (UEFI).
```
as SPL can use TF-A as well

vanmaegima commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -3,47 +3,141 @@
+
+   digraph {
+        graph [
+            label = "Secure boot flow for i.MX machines"
```
We don't define the font here, but define in the two other diagrams. For me, I prefer the font used in this one, but let's try to standardize this.

vanmaegima commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -3,47 +3,141 @@
 * ``conf/keys``: The key set is a copy of the default LmP public keys.
-* ``factory-keys``: The key set is created during the FoundriesFactory creation
-  and is unique for that Factory.
+* ``factory-keys``: The key set is created during the Factory's creation
+  and is unique for that factory.
```
I think we have a pattern on Factory, maybe @kprosise can confirm.
```suggestion
* ``factory-keys``: The key set is created during the Factory's creation
  and is unique for that Factory.
```

kprosise replied:
That's correct

vanmaegima commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -122,14 +225,35 @@ OP-TEE keys
+            -keyout privkey_ec_prime256v1.pem
+
+.. tip::
+        Remember to push the new keys to get it included in the next CI
```
```suggestion
        Remember to push the new keys to get it included them the next CI
```

vanmaegima commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -122,14 +225,35 @@ OP-TEE keys
-In order to create the key used by Linux Kernel to sign the modules a
-configuration file is needed. The `Linux Kernel documentation`_ states
-the parameters needed for the configuration file.
+A configuration file is needed to create the key used by Linux Kernel to sign the modules. The `Linux Kernel documentation`_ states
```
Are we breaking the lines at 80 columns or not?
```suggestion
A configuration file is needed to create the key used by Linux Kernel to sign the modules. The `Linux Kernel documentation`_ states
```

angolini commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -3,47 +3,141 @@
 This is used by configuring the variable ``UBOOT_SPL_SIGN_KEYNAME``.
 
 The file ``x509.genkey`` is a configuration file used for creating
-``privkey_modsign.pem`` and ``x509_modsign.crt`` which is a RSA 2048 pair in PEM
```
the "RSA 2048" information is on the table now. But the main reason to drop that is to avoid adding what RSA stands for

doanac commented:
Docs for 336be6d0aec76d4f70ea9b7ea316cf62df08feee are browsable at: https://ci.foundries.io/projects/fio-docs/builds/2115/docs/artifacts/html/index.html

doanac commented:
Docs for 3b21defbefff2fbbf58ebfbc7d7972964e2da492 are browsable at: https://ci.foundries.io/projects/fio-docs/builds/2117/docs/artifacts/html/index.html

angolini commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -122,14 +221,36 @@ OP-TEE keys
+
+.. _ref-factory-key-tfa:
+
+TrustedFirmware-A Keys
```
@kprosise, how can we ignore the vale error message for this title?

kprosise replied:
@angolini, that's fine, just ignore it! Not all vale messages are going to apply every time! If you know that it is an exception, like in this case, then all is fine!

kprosise reviewed (approved):
LGTM! @angolini  I can go ahead and merge if you're ready!

ricardosalveti commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -3,17 +3,106 @@
+        "uboot.itb"       -> "bootscr"              [label = "UBOOT_SPL_SIGN_KEYNAME"];
+        "uboot.itb"       -> "TF-A (arm64 devices)" [label = "TF_A_SIGN_KEY_PATH"];
+        "uboot.itb"       -> "OP-TEE"               [label = "OPTEE_TA_SIGN_KEY"];
+        "uboot.itb"       -> "U-Boot proper"        [label = "UBOOT_SPL_SIGN_KEYNAME"];
```
I find this a bit confusing as UBOOT_SPL_SIGN_KEYNAME is used to sign uboot.itb (so there is a direct relationship), but the key labels on the other components (coming from uboot.itb) are not related to this relationship in a way (e.g. OPTEE_TA_SIGN_KEY is used by OP-TEE when validating TAs, but here it looks like it was used by u-boot.itb when validating OP-TEE, which is not the case, since everything inside uboot.itb gets validated by UBOOT_SPL_SIGN_KEYNAME).

angolini replied:
@ricardosalveti, it makes me think that, in this diagram, OPTEE_TA_SIGN_KEY will not have an arrow associated with it. Is that right?

ricardosalveti replied:
Right, unless you add a new box for OP-TEE TAs.

angolini replied:
ok

ricardosalveti commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -3,17 +3,106 @@
+            arrowhead=none
+        ];
+        "Boot ROM"        -> "TF-A"                  [label = "RoT key"];
+        "TF-A"            -> "fip.bin"               [label = "UBOOT_SIGN_KEYNAME"];
```
Here the key is TF_A_SIGN_KEY_PATH instead (since there is no SPL).

ricardosalveti commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -3,17 +3,106 @@
+        "TF-A"            -> "fip.bin"               [label = "UBOOT_SIGN_KEYNAME"];
+        "fip.bin"         -> "bootscr"               [label = "UBOOT_SIGN_KEYNAME"];
+        "fip.bin"         -> "OP-TEE"                [label = "OPTEE_TA_SIGN_KEY"];
+        "fip.bin"         -> "U-Boot proper"         [label = "UBOOT_SIGN_KEYNAME"];
```
Same comment as on the previous boot flow, the key labels are a bit out of context here.

angolini replied:
ok

ricardosalveti commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -3,17 +3,106 @@
+        "Linux kernel"    -> "Linux kernel modules"  [label = "MODSIGN_PRIVKEY"];
+   }
+
+The following diagram shows the secure boot flow for ``intel-corei7-64`` based machines:
```
Add (booting with UEFI) since intel can have other boot methods as well.

angolini replied:
ok

ricardosalveti commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -22,28 +111,29 @@ When a FoundriesFactory is created, by default two set of keys are created under
-The **ubootdev** pair is a ``RSA`` 2048 key pair by U-Boot proper in order to validate the
-Linux® Kernel. This is used by configuring the variable ``UBOOT_SPL_SIGN_KEYNAME``.
+The **ubootdev** pair is a ``RSA`` 2048 key pair by U-Boot proper to validate the
+Linux Kernel. This is used by configuring the variable ``UBOOT_SPL_SIGN_KEYNAME``.
```
Here it should be UBOOT_SIGN_KEYNAME instead.

ricardosalveti commented on source/reference-manual/security/ff-security-summary.rst:
```diff
@@ -53,22 +59,33 @@ The exact list of keys used for the boot stack depends on the hardware used. Som
      - Owned by the customer, available as an online key for FoundriesFactory CI
      - ``MODSIGN_PRIVKEY``
+   * - :ref:`UEFI Verification Key  <ref-secure-boot-uefi>`
+     - RSA 2048 or SHA 256
```
Remove 'or SHA 256'.

ricardosalveti commented on source/reference-manual/security/ff-security-summary.rst:
```diff
@@ -53,22 +59,33 @@ The exact list of keys used for the boot stack depends on the hardware used. Som
+     - Owned by the customer, available as an online key for FoundriesFactory CI
+     - ``${UEFI_SIGN_KEYDIR}/DB.key``
+   * - :ref:`TF-A Verification Key <ref-factory-key-tfa>`
+     - RSA 2048 or SHA 256
```
We support what is used by stm32mp15, which is ECDSA (prime256v1).

angolini replied:
ok

ricardosalveti commented:
@igoropaniuk please review this one as well.

ricardosalveti commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -3,17 +3,106 @@
+        "SPL"             -> "uboot.itb"            [label = "UBOOT_SPL_SIGN_KEYNAME"];
+        "uboot.itb"       -> "bootscr"              [label = "UBOOT_SPL_SIGN_KEYNAME"];
+        "uboot.itb"       -> "TF-A (arm64 devices)" [label = "TF_A_SIGN_KEY_PATH"];
+        "uboot.itb"       -> "OP-TEE"               [label = "OPTEE_TA_SIGN_KEY"];
```
Maybe for OP-TEE you can have an additional arrow that like OP-TEE -> OP-TEE TAs, which then would show how OPTEE_TA_SIGN_KEY gets used in the end.

angolini replied:
done

doanac commented:
Docs for 0ab97e1dc875fe0546f3ae06278617b021df57d9 are browsable at: https://ci.foundries.io/projects/fio-docs/builds/2131/docs/artifacts/html/index.html

doanac commented:
Docs for 3644eee703f7bff5a0751a992f33808b30ad02cc are browsable at: https://ci.foundries.io/projects/fio-docs/builds/2138/docs/artifacts/html/index.html

angolini commented:
Do you mean, the following diff, @Tim-Anderson ?

```
-3. Run the ``gen_uefi_certs.sh`` script
+#. Install the pre-requisit packages to use ``gen_uefi_certs.sh`` script
+
+.. prompt:: bash host:~$
+
+    sudo apt install openssl, efitools, uuid-runtime
+
+#. Run the ``gen_uefi_certs.sh`` script

```

updating...

mike-scott commented on source/reference-manual/security/secure-boot-uefi.rst:
```diff
@@ -84,19 +84,25 @@ Creating UEFI Secure Boot Keys
     mkdir custom_uefi_keys_and_certs
 
-3. Run the ``gen_uefi_certs.sh`` script
+#. Install the pre-requisit packages to use ``gen_uefi_certs.sh`` script
```
Typo and missing `the`:
```
Install the prerequisite packages to use the ``gen_uefi_certs.sh`` script
```

kprosise commented:
I will do a final check after the technical reviews, so that I can review any additional content that is added.

ricardosalveti commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -122,14 +224,36 @@ OP-TEE keys
+    openssl req -new -nodes -utf8 -sha256 -days 36500 -batch -x509 \
+            -config ../conf/keys/x509.genkey -outform PEM \
+            -out x509_modsign.crt \
+            -keyout privkey_ec_prime256v1.pem
```
TF-A EC is just a key, the line is incorrect, check https://github.com/foundriesio/partner-setup/commit/fac36e2a624d117bbfe5cc380122ebc07e14622e

igoropaniuk commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -3,17 +3,109 @@
+        "fip.bin"         -> "bootscr"               [label = "UBOOT_SIGN_KEYNAME"];
+        "fip.bin"         -> "OP-TEE"                [label = "UBOOT_SIGN_KEYNAME"];
+        "OP-TEE"          -> "OP-TEE TAs"            [label = "OPTEE_TA_SIGN_KEY"];
+        "fip.bin"         -> "U-Boot proper"         [label = "UBOOT_SIGN_KEYNAME"];
```
`u-boot.itb` is missing

angolini replied:
Which are the pieces that `u-boot.itb`brings in this case?

igoropaniuk replied:
Ah sorry for confusion, I meant `u-boot.dtb`, not itb.

In both cases of containers (ITB and FIP) U-Boot DTB is a separate file, so we can mention it explicitly.

igoropaniuk reviewed (commented):
Really nice update!
I've added some cosmetic comments from my side regarding image latout

igoropaniuk commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -3,17 +3,109 @@
 ``lmp-manifest`` repository:
 
 * ``conf/keys``: The key set is a copy of the default LmP public keys.
-* ``factory-keys``: The key set is created during the FoundriesFactory creation
```
`FoundriesFactory®`

angolini commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -3,17 +3,109 @@
+        "Boot ROM"        -> "SPL"                  [label = "RoT key"];
+        "SPL"             -> "uboot.itb"            [label = "UBOOT_SPL_SIGN_KEYNAME"];
+        "uboot.itb"       -> "bootscr"              [label = "UBOOT_SPL_SIGN_KEYNAME"];
+        "uboot.itb"       -> "TF-A (arm64 devices)" [label = "UBOOT_SPL_SIGN_KEYNAME"];
```
I changed this, @igoropaniuk. Please see if you think it's readable in the final diagram

angolini commented on source/reference-manual/security/factory-keys.rst:
```diff
@@ -3,17 +3,109 @@
+        "Boot ROM"        -> "TF-A"                  [label = "RoT key"];
+        "TF-A"            -> "fip.bin"               [label = "TF_A_SIGN_KEY_PATH"];
+        "fip.bin"         -> "bootscr"               [label = "UBOOT_SIGN_KEYNAME"];
+        "fip.bin"         -> "OP-TEE"                [label = "UBOOT_SIGN_KEYNAME"];
```
Where should I add BL32 @igoropaniuk?

igoropaniuk replied:
I meant just 

```
 "fip.bin"         -> "OP-TEE (BL32)"                [label = "UBOOT_SIGN_KEYNAME"];
```

doanac commented:
Docs for ccc8f1ac4b2f5acf7741a2f5e4ac0065f1cf9892 are browsable at: https://ci.foundries.io/projects/fio-docs/builds/2162/docs/artifacts/html/index.html

angolini commented:
@igoropaniuk, I updated the doc and have a few questions for your comments.
@ricardosalveti, @Tim-Anderson, @mike-scott, I have updated the doc. Please, review if the changes are ok

Tim-Anderson commented on source/reference-manual/security/secure-boot-uefi.rst:
```diff
@@ -84,19 +84,25 @@ Creating UEFI Secure Boot Keys
     mkdir custom_uefi_keys_and_certs
 
-3. Run the ``gen_uefi_certs.sh`` script
+#. Install the prerequisit packages to use the ``gen_uefi_certs.sh`` script
```
prerequisite

Tim-Anderson commented on source/reference-manual/security/secure-boot-uefi.rst:
```diff
@@ -84,19 +84,25 @@ Creating UEFI Secure Boot Keys
 The suggested way to create a custom set of UEFI Secure Boot keys and certificates is via the `lmp-tools gen_uefi_certs.sh <https://github.com/foundriesio/lmp-tools/blob/master/security/uefi/gen_uefi_certs.sh>`_ script.
 
-1. Clone the ``lmp-tools`` repository from GitHub
+#. Clone the ``lmp-tools`` repository from GitHub
```
the auto number makes all steps 1 since there are breaks in the flow. That is why they were explicitly set. look at the converted html

Tim-Anderson reviewed (changes requested):
Please fix spelling and auto numbering.

Tim-Anderson reviewed (approved):
LGTM

kprosise commented:
@angolini It all looks good to me. Let me know when you're ready and I'll go ahead and merge

ricardosalveti reviewed (approved):
LGTM!

igoropaniuk reviewed (approved):
LGTM

kprosise commented:
@angolini ready?
